File: ios/Objective-C/GeneralSettingsObjC/GeneralSettingsObjC/Modules/BasicCell/BasicOptionalTableViewCell.m
//
//  BasicOptionalTableViewCell.m
//  GeneralSettings
//
//  Created by dynamsoft on 2021/11/22.
//

#import "BasicOptionalTableViewCell.h"

@implementation BasicOptionalTableViewCell

+ (CGFloat)cellHeight
{
    return kCellHeight;
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setupUI
{
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    
    self.titleLabel.frame = CGRectMake(kCellLeftMargin, 0, 0, kCellHeight);
    self.titleLabel.font = kFont_Regular(kCellTitleFontSize);
    self.titleLabel.textAlignment = NSTextAlignmentLeft;
    [self.contentView addSubview:self.titleLabel];
    
    self.optionalStateImageV.frame = CGRectMake(kScreenWidth - kCellRightmarin - 16 * kScreenAdaptationRadio, (kCellHeight - 16 * kScreenAdaptationRadio) / 2.0, 16 * kScreenAdaptationRadio, 16 * kScreenAdaptationRadio);
    self.optionalStateImageV.image = [UIImage imageNamed:@"checked"];
    [self.contentView addSubview:self.optionalStateImageV];
    
    self.separationLine.frame = CGRectMake(0, kCellHeight - KCellSeparationLineHeight, kScreenWidth, KCellSeparationLineHeight);
    self.separationLine.backgroundColor = kCellSeparationLineBackgroundColor;
    [self.contentView addSubview:self.separationLine];
}

/// Update UI.
- (void)updateUIWithTitle:(NSString *)titleString andOptionalState:(NSString *)optionalState
{
    self.titleLabel.text = titleString;
    self.titleLabel.width = [[ToolsHandle toolManger] calculateWidthWithText:titleString font:self.titleLabel.font AndComponentheight:self.titleLabel.height];
    
    if ([optionalState isEqualToString:@"1"]) {
        self.optionalStateImageV.hidden = NO;
    } else {
        self.optionalStateImageV.hidden = YES;
    }
}

// MARK: - Lazy loading

- (UILabel *)titleLabel
{
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
    }
    return _titleLabel;
}

- (UIImageView *)optionalStateImageV
{
    if (!_optionalStateImageV) {
        _optionalStateImageV = [[UIImageView alloc] init];
    }
    return _optionalStateImageV;
}

- (UIView *)separationLine
{
    if (!_separationLine) {
        _separationLine = [[UIView alloc] init];
    }
    return _separationLine;
}


@end

File: ios/Swift/GeneralSettingsSwift/GeneralSettingsSwift/ViewControllers/Cells/BasicOptionalTableViewCell.swift
//
//  BasicOptionalTableViewCell.swift
//  GeneralSettingsSwift
//
//  Copyright © Dynamsoft. All rights reserved.
//

import UIKit

class BasicOptionalTableViewCell: UITableViewCell {

    private lazy var titleLabel: UILabel = {
        let label = UILabel.init(frame: CGRectMake(kCellLeftMargin, 0, 0, kCellHeight))
        label.font = kFont_Regular(kCellTitleFontSize)
        label.textAlignment = .left
        return label
    }()
    
    
    lazy var optionalStateImageV: UIImageView = {
        let imageV = UIImageView.init(frame: CGRectMake(kScreenWidth - kCellRightMargin - 16, (kCellHeight - 16) / 2.0, 16, 16))
        imageV.image = UIImage(named: "checked")
        return imageV
    }()
    
    private lazy var separationLine: UIView = {
        let separationLine = UIView.init(frame: CGRectMake(0, kCellHeight - KCellSeparationLineHeight, kScreenWidth, KCellSeparationLineHeight))
        separationLine.backgroundColor = kCellSeparationLineBackgroundColor
        return separationLine
    }()
    
    static func cellHeight() -> CGFloat {
        return kCellHeight
    }
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupUI() -> Void {
        self.selectionStyle = .none
        
        self.contentView.addSubview(titleLabel)
        self.contentView.addSubview(optionalStateImageV)
        self.contentView.addSubview(separationLine)
    }
    
    func updateUI(with title: String, optionState: Bool) -> Void {
        self.titleLabel.text = title
        self.titleLabel.width = ToolsManager.shared.calculateWidth(with: title, font: self.titleLabel.font, componentHeight: self.titleLabel.height)
        self.optionalStateImageV.isHidden = !optionState
    }
}
